Sources/NIOHTTP1/HTTPDecoder.swift

```diff
@@ -15,709 +15,573 @@
-/// either the form of `HTTPClientResponsePart` or `HTTPServerRequestPart`: that is,
-/// it produces messages that correspond to the semantic units of HTTP produced by
-/// the remote peer.
-public class HTTPDecoder<HTTPMessageT>: ChannelInboundHandler, AnyHTTPDecoder {
-    public typealias InboundIn = ByteBuffer
-    public typealias InboundOut = HTTPMessageT
-
-    private let leftOverBytesStrategy: RemoveAfterUpgradeStrategy
-    private var parser = http_parser()
-    private var settings = http_parser_settings()
-
-    fileprivate var state = HTTPParserState()
-
-    fileprivate init(type: HTTPMessageT.Type, leftOverBytesStrategy: RemoveAfterUpgradeStrategy) {
-        /* this is a private init, the public versions only allow HTTPClientResponsePart and HTTPServerRequestPart */
-        assert(HTTPMessageT.self == HTTPClientResponsePart.self || HTTPMessageT.self == HTTPServerRequestPart.self)
-        self.leftOverBytesStrategy = leftOverBytesStrategy
     }
 
     deinit {
-        // Remove the stored reference to ChannelHandlerContext
-        self.parser.data = UnsafeMutableRawPointer(bitPattern: 0xdeadbee)
-
-        // Remove references to callbacks.
-        self.settings = http_parser_settings()
+        self.parser.deinitialize(count: 1)
+        self.parser.deallocate()
+        Unmanaged<BetterHTTPParser>.fromOpaque(UnsafeRawPointer(self.parser.pointee.data)).release()
     }
 
-    /// The most recent method seen by request handlers.
-    ///
-    /// Naturally, in the base case this returns nil, as servers never issue requests!
-    fileprivate func popRequestMethod() -> HTTPMethod? { return nil }
+    private func start(bytes: UnsafeRawBufferPointer, newState: HTTPDecodingState) {
+        assert(self.firstNonDiscardableOffset == nil)
+        self.firstNonDiscardableOffset = bytes.baseAddress! - self.rawBytesView.baseAddress!
+        self.decodingState = newState
+    }
 
-    private func newRequestHead(_ parser: UnsafeMutablePointer<http_parser>!) -> HTTPRequestHead {
-        let method = HTTPMethod.from(httpParserMethod: http_method(rawValue: parser.pointee.method))
-        let version = HTTPVersion(major: Int(parser.pointee.http_major), minor: Int(parser.pointee.http_minor))
-        let request = HTTPRequestHead(version: version, method: method, rawURI: state.currentURI!, headers: HTTPHeaders(buffer: cumulationBuffer!, headers: state.currentHeaders, keepAliveState: parser.keepAliveState))
-        self.state.currentHeaders.removeAll(keepingCapacity: true)
-        return request
+    private func finish(_ callout: (inout HTTPDecoderDelegate, UnsafeRawBufferPointer) -> Void) {
+        var currentFieldByteLength = 0
+        swap(&currentFieldByteLength, &self.currentFieldByteLength)
+        let start = self.rawBytesView.startIndex + self.firstNonDiscardableOffset!
+        let end = self.rawBytesView.startIndex + self.firstNonDiscardableOffset! + currentFieldByteLength
```

Comment:
I think you need a comment explaining why these are safe.

Reply:
I just added a `precondition`. They were checked in debug mode already but I think this is important enough to always check.